### ISSUE TITLE
Check for return type of overload implementation is too strict?

### ISSUE THREAD

**Issue by ilevkivskyi** [python/mypy#7338]
Consider this simplified example:
```python
from typing import overload, Optional, List
  
@overload
def f(x: int) -> List[int]: ...
@overload
def f(x: None) -> List[Optional[int]]: ...

def f(x: Optional[int]) -> List[Optional[int]]:
    return [x]
```
Mypy currently complains about it that:
```
Overloaded function implementation cannot produce return type of signature 1
```
Although I understand why this happens, it is a bit annoying and can be perceived as a false positive. Maybe we should use `is_overlapping_types()` instead of `is_subtype()` for the check?

cc @Michael0x2a

**Comment by ilevkivskyi** [comment]
That part of the PEP is incorrect should be fixed. Body of the implementation can still be type-checked as any other function.
